Generate correlated RDPG graph pairs

SHAAAAN opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/neurodata/graspy/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Closes #310 
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

#### Generate correlated Random Dot Product Graph (RDPG) graph pairs

RDPG is a latent position generative model, in which the probability of an edge existing between pairs of vertices is determined by the dot product of the associated latent position vectors. 

The principle is that given a matrix X, the probability matrix will be calculated as P = XX^T, and then the correlated graph pairs will be generate based on the P and correlation. 

Here is the link to the paper: https://www.tandfonline.com/doi/pdf/10.1080/01621459.2012.699795

bdpedigo commented:
Hi @SHAAAAN, please request a peer review first

SHAAAAN commented:
Here is a notebook to show its performance: 
https://nbviewer.jupyter.org/github/SHAAAAN/MyGraspy/blob/master/notebook_rdpg_corr.ipynb

bdpedigo commented:
Hi @SHAAAAN did you ever get someone to peer review this?

SHAAAAN commented:
> Hi @SHAAAAN did you ever get someone to peer review this?

Hi @bdpedigo, I have invited Ali to do it last week. He agreed to do it, but I'm not sure when he has time to do it.

asaadeldin11 commented:
@SHAAAAN  I looked over your code and everything mostly looks good. Main comments would be make the description to the PR more descriptive. What does RDPG stand for? Why would someone want to use it as opposed to some other type of simulated graph pair? Also cite the paper that your implementation is based.
An additional comment I had is based on the netifly preview. Under the tutorial section, a subsection of your tutorial is somehow a section in the sidebar. Not sure how this happened but try to resolve it.
Other than these small comments, I'd say looks good
<img width="300" alt="Screen Shot 2020-02-27 at 12 45 43 PM" src="https://user-images.githubusercontent.com/54996865/75470740-1fe3a800-595f-11ea-9531-dc0bb1f56072.png">

SHAAAAN commented:
> @SHAAAAN I looked over your code and everything mostly looks good. Main comments would be make the description to the PR more descriptive. What does RDPG stand for? Why would someone want to use it as opposed to some other type of simulated graph pair? Also cite the paper that your implementation is based.
> An additional comment I had is based on the netifly preview. Under the tutorial section, a subsection of your tutorial is somehow a section in the sidebar. Not sure how this happened but try to resolve it.
> Other than these small comments, I'd say looks good
> <img alt="Screen Shot 2020-02-27 at 12 45 43 PM" width="300" src="https://user-images.githubusercontent.com/54996865/75470740-1fe3a800-595f-11ea-9531-dc0bb1f56072.png">

Hi @asaadeldin11 , thanks for your review. I have added some description of this PR and fixed the error in the tutorial section. Does it look better now?

bdpedigo commented:
Hi @SHAAAAN I don't understand how you are calculating correlation here. Remember how for SBM we had to condition on P in order to accurately calculate the correlation? How do you do that here?

bdpedigo commented:
I think in this case it could make more sense to do: 
- sample `n_sims` pairs where `n_sims` is like 200-500
- do not compute correlation between pairs of graphs directly, like we have been doing 
- for each {i, j} pair (edge) you now have 200-500 samples for graph `A` and just as many for graph `B`
- for each {i, j} pair, treat the sampled edges from `A` as a vector (`e_A`), and same for `B` (`e_B`). Take correlation between `e_A` and `e_B`. Do this for all {i, j} pairs 
- take the average of the above, output that as the overall correlation. 

My logic for this is that in the RDPG case, all elements of P are potentially unique. So to condition on P, we have to do it by edge (I think).

Make sense? agree/disagree?

SHAAAAN commented:
> I think in this case it could make more sense to do:
> 
> * sample `n_sims` pairs where `n_sims` is like 200-500
> * do not compute correlation between pairs of graphs directly, like we have been doing
> * for each {i, j} pair (edge) you now have 200-500 samples for graph `A` and just as many for graph `B`
> * for each {i, j} pair, treat the sampled edges from `A` as a vector (`e_A`), and same for `B` (`e_B`). Take correlation between `e_A` and `e_B`. Do this for all {i, j} pairs
> * take the average of the above, output that as the overall correlation.
> 
> My logic for this is that in the RDPG case, all elements of P are potentially unique. So to condition on P, we have to do it by edge (I think).
> 
> Make sense? agree/disagree?

Agree. I'll test it in this way and see what's happening.

SHAAAAN commented:
> > I think in this case it could make more sense to do:
> > 
> > * sample `n_sims` pairs where `n_sims` is like 200-500
> > * do not compute correlation between pairs of graphs directly, like we have been doing
> > * for each {i, j} pair (edge) you now have 200-500 samples for graph `A` and just as many for graph `B`
> > * for each {i, j} pair, treat the sampled edges from `A` as a vector (`e_A`), and same for `B` (`e_B`). Take correlation between `e_A` and `e_B`. Do this for all {i, j} pairs
> > * take the average of the above, output that as the overall correlation.
> > 
> > My logic for this is that in the RDPG case, all elements of P are potentially unique. So to condition on P, we have to do it by edge (I think).
> > Make sense? agree/disagree?
> 
> Agree. I'll test it in this way and see what's happening.

Hi @bdpedigo , I update the notebook and the link is : https://nbviewer.jupyter.org/github/SHAAAAN/MyGraspy/blob/master/notebook_rdpg_corr_update.ipynb
Do you think it make sense?

bdpedigo commented:
Hi @SHAAAAN, a few comments: 
- I don't think you need to use `np.where` in your `cal_rho` function, you can just index with a boolean array. Only reason I mention is cause `np.where` can be slow and IDK how long this notebook takes to run
- In general, don't use `np.mat`, use `np.array`
- Why not run all of the cal_rhos in a for loop?
- The last plot is no longer informative with this many points. But I am confused because it also doesn't seem to match the other plots. for example, looking at the first column of dots at -0.6, they don't seem even around 0. Try making the alpha value small. or maybe this plot is no longer meaningful. 
- Do you know why the correlation values look more discreet sometimes? (e.g. for rho=0.9, there are a few really dense bars)

SHAAAAN commented:
> * you can just index with a boolean array.

Thanks for your advice since the running times of these functions now are really long.

Here the last plot shows the difference between calculated rhos and expected rhos, so they are all around 0. Maybe there are too many data points in the last plot so that the differences range wide here. I can change the alpha smaller or just delete that plot.

And I am also confused about the discreet bars. And the more rho closes to 1 or -1, the more discreet the bars are.

I'll optimize it a little bit and see whether there are some changes.

SHAAAAN commented:
Hi @bdpedigo , I have optimize the functions and here is the new notebook:
https://nbviewer.jupyter.org/github/SHAAAAN/MyGraspy/blob/master/codes/notebook_rdpg_corr_update.ipynb

I have tried 100 and 1000 pairs simulations when rho=0.9, it seems that as pairs become larger, the calculated rhos are more centered. But there are still some discreet bars in the distplot. 
Here is the plot with 1000 pairs when rho=0.9:
![p1000](https://user-images.githubusercontent.com/45969738/76358735-5242bd00-62f0-11ea-9f59-b7682342d3a7.png)

And I checked the plots in corr_sbm property notebook, the plots look similar:
![-0 99](https://user-images.githubusercontent.com/45969738/76358375-bc0e9700-62ef-11ea-8d09-467b81bee794.png)
![0 99](https://user-images.githubusercontent.com/45969738/76358382-be70f100-62ef-11ea-8f19-2988814a610d.png)

bdpedigo commented on graspy/simulations/rdpg_corr.py:
```diff
@@ -0,0 +1,114 @@
+        
+    References
+    ----------
+    .. [1] Sussman, D.L., Tang, M., Fishkind, D.E., Priebe, C.E.  "A
```
@SHAAAAN update this reference

SHAAAAN replied:
Hi Pedigo, can you tell me which reference can I refer here?

bdpedigo replied:
just use this for now https://dl.acm.org/doi/10.5555/2627435.2750357

bdpedigo replied:
but add the actual reference, not just the link, please

bdpedigo commented on graspy/simulations/rdpg_corr.py:
```diff
@@ -0,0 +1,114 @@
+from graspy.simulations import p_from_latent, sample_edges_corr
+
+
+def rdpg_corr(X, Y, r, rescale, directed, loops):
```
@SHAAAAN some of these should be kwargs, like normal

SHAAAAN replied:
I don't understand what do you mean here, do you mean adding the default values?

bdpedigo replied:
yes, these are called kwargs (keyword arguments)

bdpedigo commented on tests/test_rdpg_corr.py:
```diff
@@ -0,0 +1,49 @@
+            self.X, self.Y, self.r, rescale=False, directed=False, loops=False
+        )
+
+        # check rho
```
@SHAAAAN add a test to check P makes sense as well.

bdpedigo replied:
@SHAAAAN can you point me to where this is addressed?

bdpedigo replied:
@SHAAAAN what i meant was, a test demonstrating that the output values of P after sampling graphs are close to those those that we expect based on the input matrices. Does that make sense? I think you are talking about input checks, which is good, but not what I am asking about.

bdpedigo replied:
as in, you assert that the correlation is correct, but you don't assert that P is correct

SHAAAAN replied:
I see what you mean now. Can I just copy the code from the test in rdpg for p?

bdpedigo commented on tests/test_rdpg_corr.py:
```diff
@@ -0,0 +1,49 @@
+        with self.assertRaises(ValueError):
+            sample_edges(x2)  # wrong shape for P
+
+    def test_rdpg_corr(self):
```
@SHAAAAN there are no random seeds on your tests so they could fail randomly

bdpedigo commented on tests/test_rdpg_corr.py:
```diff
@@ -0,0 +1,49 @@
+import unittest
+from graspy.simulations.simulations import sample_edges, rdpg, p_from_latent
```
@SHAAAAN remove any imports in this header that you don't use

bdpedigo commented:
@SHAAAAN for the RDPG corr tutorial, don't just copy/paste from the other tutorial, I would say just link to it and mention that an explanation of the uncorrelated model is there.

bdpedigo commented:
@SHAAAAN also in tutorial, you don't explain anything about the correlated model, or what you mean by `r`

bdpedigo commented:
@SHAAAAN I think the tutorial should explain the model more. I would do something like what is [here ](https://graspy.neurodata.io/tutorials/matching/sgm.html).

Steal the code for sampling an SBM pair, plotting the two graphs, and showing the difference. I would also print out the number of edges that are different between the two graphs. Then, for several values of rho, just show that plot with the 3 columns (graph 1, graph 2, diff). say rho = negative, rho = 0.3ish, rho=0.9ish. this way people will see that as you increase rho the number of edges that are the same goes up.

bdpedigo commented:
@SHAAAAN when you are finished with the above, make sure to clear the notebook output in jupyter before uploading. in jupyter there should be a button for "clear all output" somewhere, just do this, save, push

bdpedigo commented on graspy/simulations/rdpg_corr.py:
```diff
@@ -0,0 +1,114 @@
+    --------
+    >>> np.random.seed(1234)
+    
+    To gnerate random latent positions using 2-dimensional Dirichlet distribution:
```
typo (generate)

SHAAAAN commented:
Hi @bdpedigo, I have made some revision based on your comments. Please tell me if there is other thing that I need to change.

SHAAAAN commented:
> @SHAAAAN I think the tutorial should explain the model more. I would do something like what is [here ](https://graspy.neurodata.io/tutorials/matching/sgm.html).
> 
> Steal the code for sampling an SBM pair, plotting the two graphs, and showing the difference. I would also print out the number of edges that are different between the two graphs. Then, for several values of rho, just show that plot with the 3 columns (graph 1, graph 2, diff). say rho = negative, rho = 0.3ish, rho=0.9ish. this way people will see that as you increase rho the number of edges that are the same goes up.

I just added a plot of difference rate with different correlations from -0.5 to 0.9 in the tutorial:
![Screen Shot 2020-03-25 at 1 46 53 PM](https://user-images.githubusercontent.com/45969738/77568729-73cea780-6e9f-11ea-84cb-dd850ae3ea4c.png)

bdpedigo commented:
Cool, what is "difference rate"?

SHAAAAN commented:
> Cool, what is "difference rate"?

Its the difference between G1 and G2 with different correlation values

bdpedigo commented:
> “r” is the correlation between the graph pair, which should be (-1,1) (the value could be more limited in practice due to the value of “r”).

this doesn't make sense, due to the value of P? I would explicitly say that not all combinations of r and p are possible.

bdpedigo commented:
For the tutorial notebook, please just use the code from the SBM/SGM tutorial that I linked to make the 3 heatmaps (G1, G2, and diff) all show up on the same axes, and use the right colormapping so that you can see positive and negative differences.

bdpedigo commented:
You don't explain "difference rate" in the notebook and this term is confusing to me. Also don't hard code values like `np.sum(abs(g1-g2))/(100*100)` because if someone changes the size of the graph, for example, then dividing by 100*100 will not make sense anymore

SHAAAAN commented on tests/test_rdpg_corr.py:
```diff
@@ -0,0 +1,58 @@
+        with self.assertRaises(ValueError):
+            sample_edges(x2)  # wrong shape for P
+
+        if any(self.X[self.X > 1]) or any(self.X[self.X < -1]):  # wrong values for P
```
I check the input values of X here since P is an intermediate variable

SHAAAAN commented on tests/test_rdpg_corr.py:
```diff
@@ -0,0 +1,58 @@
+            self.X, self.Y, self.r, rescale=False, directed=False, loops=False
+        )
+
+        # check the dimention of g1, g2
```
And I check the dimension of g1 and g2 here to make sure they have the same dimension with P indirectly.

bdpedigo commented on tests/test_rdpg_corr.py:
```diff
@@ -0,0 +1,73 @@
+        graphs = []
+        P = p_from_latent(X, rescale=True, loops=True)
+        for i in range(1000):
+            graphs.append(sample_edges(P, directed=True, loops=True))
```
@SHAAAAN you aren't even testing your own function?
